### ISSUE TITLE
Memory Leak?

### ISSUE THREAD

**Issue by Buzzunu** [neslib/Neslib.Xml#13]
Thank you very much for this wonderful library, it has been of great use to me.
One detail that I still haven't managed to solve, I installed MadExcept (https://www.madshi.net/madExceptDescription.htm) and noticed that a memory leak was appearing (VirtualAlloc), as using “ReportMemoryLeaksOnShutdown” no memory leak appears, I contacted the author of MadExcept to check if it was a false positive, but according to him:

> "looks like true leaks to me.
> "ReportMemoryLeaksOnShutdown" does only one thing: It looks for leaks in Delphi's own memory manager. madExcept's leak reporting is *dramatically* more extensive. It hooks around 500 win32 APIs and notifies you about all kinds of leaks, including OS memory manager leaks, leaked window handles, thread handles, process handles, GDI handles etc etc.
> Looking at your leak report, none of the leaks in there are leaks that "ReportMemoryLeaksOnShutdown" even looks for." madshi.

So I created a test project where I just read the nasa.xml file from the example and you can see that the memory increases and doesn't decrease anymore.
I don't have much experience in coding directly in memory, is there any possibility of looking into this issue?
I've uploaded the test project and the memory leak log to my onedrive in case it's useful.
[Project sample](https://1drv.ms/u/s!AvDBlgi36foehckdniL_YD1-a19yKg?e=DUG02t)
[Log](https://1drv.ms/u/s!AvDBlgi36foehckcxizeF5rn7Qhw5A?e=cdUAqm)

**Comment by neslib** [comment]
Hi Jonathas,

 

Thank you for this report. I will try to find some time this weekend to look into it.

 

Regards,

Erik

 

From: Jonathas Paiva ***@***.***> 
Sent: Thursday, September 19, 2024 2:45 PM
To: neslib/Neslib.Xml ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [neslib/Neslib.Xml] Memory Leak? (Issue #13)

 

Thank you very much for this wonderful library, it has been of great use to me.
One detail that I still haven't managed to solve, I installed MadExcept (https://www.madshi.net/madExceptDescription.htm) and noticed that a memory leak was appearing (VirtualAlloc), as using “ReportMemoryLeaksOnShutdown” no memory leak appears, I contacted the author of MadExcept to check if it was a false positive, but according to him:

"looks like true leaks to me.
"ReportMemoryLeaksOnShutdown" does only one thing: It looks for leaks in Delphi's own memory manager. madExcept's leak reporting is dramatically more extensive. It hooks around 500 win32 APIs and notifies you about all kinds of leaks, including OS memory manager leaks, leaked window handles, thread handles, process handles, GDI handles etc etc.
Looking at your leak report, none of the leaks in there are leaks that "ReportMemoryLeaksOnShutdown" even looks for." madshi.

So I created a test project where I just read the nasa.xml file from the example and you can see that the memory increases and doesn't decrease anymore.
I don't have much experience in coding directly in memory, is there any possibility of looking into this issue?
I've uploaded the test project and the memory leak log to my onedrive in case it's useful.
Project sample <https://1drv.ms/u/s!AvDBlgi36foehckdniL_YD1-a19yKg?e=DUG02t> 
Log <https://1drv.ms/u/s!AvDBlgi36foehckcxizeF5rn7Qhw5A?e=cdUAqm>

**Comment by neslib** [comment]
Hi Jonathas,

 

I don’t have madExcept so I cannot test with that. But I tested your sample project as well as a sample of my own, and I don’t see any continuous memory increase. I’ve uses both Windows Task Manager and Process Explorer (https://learn.microsoft.com/en-us/sysinternals/downloads/process-explorer) to check memory usage.

 

When loading the document for the first time, memory increased a little because of some memory that stays alive for the duration of the process. But after that, every time you re-load the document, the memory usage goes up and down the same amount. I used this little console project to test this:

 

  for var I := 0 to 9 do

  begin

    WriteLn('Before loading. Press [Enter]...');

    ReadLn;

 

    var Document := TXmlDocument.Create;

    Document.Load(‘nasa.xml');

 

    WriteLn('After loading, before release. Press [Enter]...');

    ReadLn;

 

    Document := nil;

    WriteLn('After release. Press [Enter]...');

    ReadLn;

  end;

 

Every time the document is loaded, memory usage goes up by about 40 MB. But when the document is released, memory goes down by the same 40 MB again.

I also counted every time VirtualAlloc(MEM_RESERVE) and VirtualFree is called, and they are both called exactly the same number of times (104 times for the nasa.xml file in 32-bit mode), meaning that all memory allocated using VirtualAlloc is also freed again using VirtualFree.

 

Maybe there is another issue somewhere? 

 

Hope this helps,

Erik van Bilsen

 

From: Jonathas Paiva ***@***.***> 
Sent: Thursday, September 19, 2024 2:45 PM
To: neslib/Neslib.Xml ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [neslib/Neslib.Xml] Memory Leak? (Issue #13)

 

Thank you very much for this wonderful library, it has been of great use to me.
One detail that I still haven't managed to solve, I installed MadExcept (https://www.madshi.net/madExceptDescription.htm) and noticed that a memory leak was appearing (VirtualAlloc), as using “ReportMemoryLeaksOnShutdown” no memory leak appears, I contacted the author of MadExcept to check if it was a false positive, but according to him:

"looks like true leaks to me.
"ReportMemoryLeaksOnShutdown" does only one thing: It looks for leaks in Delphi's own memory manager. madExcept's leak reporting is dramatically more extensive. It hooks around 500 win32 APIs and notifies you about all kinds of leaks, including OS memory manager leaks, leaked window handles, thread handles, process handles, GDI handles etc etc.
Looking at your leak report, none of the leaks in there are leaks that "ReportMemoryLeaksOnShutdown" even looks for." madshi.

So I created a test project where I just read the nasa.xml file from the example and you can see that the memory increases and doesn't decrease anymore.
I don't have much experience in coding directly in memory, is there any possibility of looking into this issue?
I've uploaded the test project and the memory leak log to my onedrive in case it's useful.
Project sample <https://1drv.ms/u/s!AvDBlgi36foehckdniL_YD1-a19yKg?e=DUG02t> 
Log <https://1drv.ms/u/s!AvDBlgi36foehckcxizeF5rn7Qhw5A?e=cdUAqm>

**Comment by Buzzunu** [comment]
Hello Erick, thank you very much for the quick reply.

Sorry the initial comment on the issue didn't contain much information.
Regarding the MadExcept tool, it's free for non-commercial use or you can use a trial, although I don't know if it's a real impediment.
But when you use MadExcept in a debugging environment with an active memory leak detector, ram consumption rises more than normal precisely because of the tool's internal workings, and in fact, in a production environment (release) the increase in consumption is small (3mb on average).
Strangely, there is no continuous increase in memory, only a “fixed” increase the first time it is used, which MadExcept says is a VirtualAlloc type from line 2137.

I'm attaching 3 gifs showing the memory consumption, all based on the example you provided.
The first without madexcept.
The second with madexcept.
The third with madexcept and showing the log afterwards.

![Neslib_Leak](https://github.com/user-attachments/assets/8c0f1daa-a1fc-475b-b211-2862f9fcd4b9)
![Neslib_Leak_MadExcept](https://github.com/user-attachments/assets/4f21159a-4807-437f-949c-415faa2249af)
![Neslib_Leak_MadExcept2](https://github.com/user-attachments/assets/b15d3870-4fdd-4725-a77e-9b76f007384a)

I came across this question because we're trying to zero out memory leaks in a project, and memory consumption in itself doesn't create a problem.

**Comment by Buzzunu** [comment]
Good evening.
I took some time out this evening to run some tests.
The leak only occurs if the file is large enough to have to allocate 2 or more blocks of memory.
I took an XML in which it passes through the “AllocBlock” function 4 times, however it reserves (MEM_RESERVE) 1 time, commits (MEM_COMMIT) 4 times and gives free(VirtualFree/MEM_RELEASE) 1 time.
At the end of the project, MadExcept generates a log with a memory leak of 3 VirtualAlloc.
I did a test as follows:
I created 3 variables in TXmlDocument:
FMemoryReserved: Integer;
FMemoryCommit: Integer;
FMemoryFree: Integer;

Where I increment them when I pass each time I call AllocBlock to reserve or commit, and also increment FMemoryFree in Clear when I call the VirtualFree method.
At the end I display the evetns with “OutputDebugString”.
If the XML is small and passes only once to reserve and commit, there is no memory leak.

I honestly don't know if it's a false positive or not, but that's the scenario I managed to get to at the moment.

**Comment by neslib** [comment]
Hi Jonathas,

 

Thank you for these tests. I’ve reviewed my code and the documentation on VirtualAlloc and VirtualFree and I think this may be  a false positive.

 

Because of the flags I use in these calls, the behavior you mentioned is as expected. In this case, VirtualAlloc is called once with the MEM_RESERVE flag to reserve a large block of memory (usually 64 KB). Then calls VirtualAlloc with the MEM_COMMIT flag to actually allocate 4 pages of memory (usually 4 KB each). Then VirtualFree is called once to free the entire large 64 KB block, which includes these 4 smaller pages.

 

So I think the log incorrectly shows that the leak is caused by calls to VirtualAlloc. Because you don’t have to match every VirtualAlloc call with a VirtualFree call (as you normally would with GetMem and FreeMem). Maybe madExcept is expecting that every VirtualAlloc(MEM_COMMIT) has a matching VirtualFree call, which is not always necessary (it depends on the flags being used.).

 

Hope this makes sense,

Erik

**Comment by Buzzunu** [comment]
> Hi Jonathas,
Thank you for these tests. I’ve reviewed my code and the documentation on VirtualAlloc and VirtualFree and I think this may be  a false positive.
Because of the flags I use in these calls, the behavior you mentioned is as expected. In this case, VirtualAlloc is called once with the MEM_RESERVE flag to reserve a large block of memory (usually 64 KB). Then calls VirtualAlloc with the MEM_COMMIT flag to actually allocate 4 pages of memory (usually 4 KB each). Then VirtualFree is called once to free the entire large 64 KB block, which includes these 4 smaller pages.
So I think the log incorrectly shows that the leak is caused by calls to VirtualAlloc. Because you don’t have to match every VirtualAlloc call with a VirtualFree call (as you normally would with GetMem and FreeMem). Maybe madExcept is expecting that every VirtualAlloc(MEM_COMMIT) has a matching VirtualFree call, which is not always necessary (it depends on the flags being used.).
Hope this makes sense,
Erik 

Good morning, Erik.
Thanks for the feedback, after your message and debugging the code I created a simple example to simulate the “problem” and together with the creator of MadExcept it was confirmed that it was a false positive along with the promise of a fix.
So for anyone else who encounters this situation, just wait until a fix comes out soon.
Thanks again.

**Comment by neslib** [comment]
Hi Jonathas,

 

I’m glad that you got to the bottom of this issue and that you were able to help Madshi improve their product. And I’m glad there is no memory leak in my library too.

 

Regards,

Erik
